src/filters/matches.rs

```diff
@@ -0,0 +1,439 @@
+        Ok(Self {
+            on_read: value
+                .on_read
+                .map(proto::matches::DirectionalConfig::try_into)
+                .transpose()
+                .map_err(|error: eyre::Report| {
+                    ConvertProtoConfigError::new(error, Some("on_read".into()))
+                })?,
+            on_write: value
+                .on_write
+                .map(proto::matches::DirectionalConfig::try_into)
+                .transpose()
+                .map_err(|error: eyre::Report| {
+                    ConvertProtoConfigError::new(error, Some("on_write".into()))
+                })?,
+        })
+    }
+}
+
+impl TryFrom<proto::matches::DirectionalConfig> for DirectionalConfig {
+    type Error = eyre::Report;
+
+    fn try_from(value: proto::matches::DirectionalConfig) -> Result<Self, Self::Error> {
+        Ok(Self {
+            metadata_key: value.metadata_key.ok_or_else(|| {
+                ConvertProtoConfigError::new("Missing", Some("metadata_key".into()))
+            })?,
+            branches: value
+                .branches
+                .into_iter()
+                .map(proto::matches::Branch::try_into)
+                .collect::<Result<_, _>>()?,
+            fallthrough: value
+                .fallthrough
+                .ok_or_else(|| ConvertProtoConfigError::new("Missing", Some("fallthrough".into())))?
+                .try_into()?,
+        })
+    }
+}
+
+/// Configuration for a specific direction
+#[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+pub struct DirectionalConfig {
+    #[serde(rename = "metadataKey")]
+    pub metadata_key: String,
+    pub branches: Vec<Branch>,
+    #[serde(default)]
+    pub fallthrough: Fallthrough,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+pub struct Branch {
+    pub value: crate::metadata::Value,
+    pub filter: String,
+    pub config: Option<ConfigType>,
+}
+
+impl TryFrom<proto::matches::Branch> for Branch {
+    type Error = eyre::Report;
+
+    fn try_from(branch: proto::matches::Branch) -> Result<Self, Self::Error> {
+        Ok(Self {
+            value: branch
+                .value
+                .ok_or_else(|| ConvertProtoConfigError::new("Missing", Some("value".into())))?
+                .try_into()?,
+            filter: branch
+                .filter
+                .ok_or_else(|| ConvertProtoConfigError::new("Missing", Some("filter".into())))?,
+            config: branch.config.map(ConfigType::Dynamic),
+        })
+    }
+}
+
+///  How the [`Matches`] filter should handle no branch being matched.
+#[derive(Debug, PartialEq)]
+pub enum Fallthrough {
+    /// The packet will be passed onto the next filter.
+    Pass,
+    /// The packet will be dropped. **Default behaviour**
+    Drop,
+    /// The filter specified in `filter` will be called.
+    Filter {
+        filter: String,
+        config: Option<ConfigType>,
```

Comment:
I think this can be Option<serde_yaml::Value> rather than a full config type or? since this is only used for the yaml config? in which case it probably allows us to avoid the custom serde we have for fallthrough? (same for Branch)

Reply:
It is needed since someone can send a `Fallthrough::Filter` through xDS/protobuf, not just in the static configuration

Reply:
Ah so iiuc its not necessarily it being used through xds but rather that to use [deserialize](https://github.com/googleforgames/quilkin/pull/447/files/c5f4bbe10f1aaea625378d3585d33d61ffca6230#diff-19529b6c5a2c25b8fe5e26633ff818cc2a563656df8c69745b3429ed31d0b937R167-R169) we need to be able to convert from proto to yaml and there isn't a reasonable way to do that for the transparent filter config field? We don't have to use `deserialize` since its only a helper function, we can call the convert function ourselves and instantiate the filter i.e 
```rust
fn create_filter(&self, args: CreateFilterArgs) -> Result<FilterInstance, Error> {
  let (config_json, on_read, on_write) = match args {
    ConfigType::Static(config) => {
      // Instantiate sub filters from static config
    },
    ConfigType::Dynamic(config) => {
      // Instantiate sub filters from dynamic config
    }
  }

  let filter = Matches::new(on_read, on_write)?;
  Ok(FilterInstance::new(
    config_json,
    Box::new(filter) as Box<dyn Filter>,
  ))
}
```